### PR TITLE
Fix: set current shader null when it is destroyed.

### PR DESCRIPTION
**Pull request by orange4glace**
##### Description of change
`ShaderSystem` should set its `shader` when it is destroyed.

```html
<script src="./pixi.js"></script>


<body></body>
<script>
  
  const vert1 = `
  #version 300 es
  in vec4 aPos;
  void main(void) {
    gl_PointSize = 20.;
    gl_Position = vec4(aPos.x + 0.005, aPos.yzw);
  }
  `
  
  const frag1 = `
  #version 300 es
  precision mediump float;
  out vec4 fragColor;
  void main(void){
    fragColor = vec4(1.0, 0.0, 0.0, 1.0);
  }
  `
  
  const program = new PIXI.Program(vert1, frag1);
  const shader = new PIXI.Shader(program);

  const geometry = new PIXI.Geometry()
    .addAttribute('aPos', [-1, 0, 0, 1]);
  const mesh = new PIXI.Mesh(geometry, shader, undefined, PIXI.DRAW_MODES.POINTS);

  const container = new PIXI.Container();
  container.addChild(mesh);
  
  // Create the application helper and add its render target to the page
  let app = new PIXI.Application({ width: 640, height: 360 });
  document.body.appendChild(app.view);
  
  app.stage.addChild(container);

  setTimeout(() => {
    container.removeChild(mesh);
    shader.destroy();
    mesh.destroy();
  }, 2000)
  
  let elapsed = 0.0;
  app.ticker.add((delta) => {
    elapsed += delta;
  });
  

  
</script>
```

In the above code, 

```ts
  setTimeout(() => {
    container.removeChild(mesh);
    shader.destroy();
    mesh.destroy();
  }, 2000)
```

There should be no case that `shader` is accessed inside the PIXI system because `mesh` is removed from the scene.
But destroying shader does not set `ShaderSystem.shader` as null and `ProjectionSystem` trying to update `renderer.shader.shader` in `update()` which is already destroyed so `uniforms` is null, ends up with an error.

```
core.mjs:4028 Uncaught TypeError: Cannot read properties of null (reading 'uniforms')
    at get uniforms [as uniforms] (core.mjs:4028:30)
    at ProjectionSystem.update (core.mjs:4666:63)
    at RenderTextureSystem.bind (core.mjs:4746:30)
    at ObjectRendererSystem.render (core.mjs:6870:28)
    at _Renderer.render (core.mjs:6125:25)
    at _Application.render (app.mjs:23:19)
    at TickerListener.emit (ticker.mjs:37:17)
    at Ticker.update (ticker.mjs:216:29)
    at Ticker._tick (ticker.mjs:94:14)
```

I think this needs to be fixed in 6.x also.

##### Pre-Merge Checklist

- [ ] Tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] Lint process passed (`npm run lint`)
- [ ] Tests passed (`npm run test`)

### CODE REVIEW

File: packages/core/src/shader/Shader.ts [hunk 1]
```diff
@@ -70,8 +75,16 @@ export class Shader
...
 
+    /** Disposes WebGL resources that are connected to this shader. */
+    dispose(): void
+    {
+        this.disposeRunner.emit(this, false);
+    }
+
     destroy(): void
     {
+        this.dispose();
```
Comment (bigtimebuddy) [inline packages/core/src/shader/Shader.ts:86]: You probably don't need the dispose function. And it's kind of confusing to have destroy and dispose anyways. Maybe just in-line the runner emit here?

**Review by bigtimebuddy — COMMENTED** [review]
Good catch! Is it possible to make a small test that recreates this error?

**Comment by orange4glace** [comment]
> Good catch! Is it possible to make a small test that recreates this error?

Added!

File: packages/core/src/shader/Shader.ts [hunk 2]
```diff
@@ -72,6 +77,8 @@ export class Shader
 
     destroy(): void
     {
+        this.disposeRunner.emit(this, false);
```
Comment (bigtimebuddy) [inline packages/core/src/shader/Shader.ts:80]: Shouldn't we dispose the Runner too?

File: packages/core/src/shader/ShaderSystem.ts [hunk 3]
```diff
@@ -94,6 +94,8 @@ export class ShaderSystem implements ISystem
      */
     bind(shader: Shader, dontSync?: boolean): GLProgram
     {
+        shader.disposeRunner.add(this);
```
[the remark this answers is not in the document]
Comment (orange4glace) [reply packages/core/src/shader/ShaderSystem.ts:97]: I thought there's a logic for preventing adding duplicated listener. :?
https://github.com/pixijs/pixijs/blob/dev/packages/runner/src/Runner.ts#L119

[the remark this answers is not in the document]
Comment (bigtimebuddy) [reply packages/core/src/shader/ShaderSystem.ts:97]: It's not the duplicate listeners thats the problem, is the having every shader have a listener to dispose, when really, we only need the active shader to do that. Wdyt?

**Review by bigtimebuddy — APPROVED** [review]

**Comment by bigtimebuddy** [comment]
Thanks!

File: packages/core/src/shader/ShaderSystem.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (orange4glace) [reply packages/core/src/shader/ShaderSystem.ts:97]: > It's not the duplicate listeners thats the problem, is the having every shader have a listener to dispose, when really, we only need the active shader to do that. Wdyt?

Sorry for the late response. I was busy meanwhile 😬.
That's very reasonable! Maybe we can fix that later :)
I think either way does not huge impact in a performance though (and you've already merged it 😄)

**Pull request merged**